google/resource_cloudiot_registry.go

```diff
@@ -0,0 +1,381 @@
+			PubsubTopicName: v.(string),
+		}
+	}
+	return nil
+}
+
+func buildStateNotificationConfig(config map[string]interface{}) *cloudiot.StateNotificationConfig {
+	if v, ok := config["pubsub_topic_name"]; ok {
+		return &cloudiot.StateNotificationConfig{
+			PubsubTopicName: v.(string),
+		}
+	}
+	return nil
+}
+
+func buildMqttConfig(config map[string]interface{}) *cloudiot.MqttConfig {
+	if v, ok := config["mqtt_enabled_state"]; ok {
+		return &cloudiot.MqttConfig{
+			MqttEnabledState: v.(string),
+		}
+	}
+	return nil
+}
+
+func buildHttpConfig(config map[string]interface{}) *cloudiot.HttpConfig {
+	if v, ok := config["http_enabled_state"]; ok {
+		return &cloudiot.HttpConfig{
+			HttpEnabledState: v.(string),
+		}
+	}
+	return nil
+}
+
+func buildPublicKeyCertificate(certificate map[string]interface{}) *cloudiot.PublicKeyCertificate {
+	cert := &cloudiot.PublicKeyCertificate{
+		Format:      certificate["format"].(string),
+		Certificate: certificate["certificate"].(string),
+	}
+	return cert
+}
+
+func expandCredentials(credentials []interface{}) []*cloudiot.RegistryCredential {
+	certificates := make([]*cloudiot.RegistryCredential, len(credentials))
+	for i, raw := range credentials {
+		cred := raw.(map[string]interface{})
+		certificates[i] = &cloudiot.RegistryCredential{
+			PublicKeyCertificate: buildPublicKeyCertificate(cred["public_key_certificate"].(map[string]interface{})),
+		}
+	}
+	return certificates
+}
+
+func createDeviceRegistry(d *schema.ResourceData) *cloudiot.DeviceRegistry {
+	deviceRegistry := &cloudiot.DeviceRegistry{}
+	if v, ok := d.GetOk("event_notification_config"); ok {
+		deviceRegistry.EventNotificationConfigs = make([]*cloudiot.EventNotificationConfig, 1, 1)
+		deviceRegistry.EventNotificationConfigs[0] = buildEventNotificationConfig(v.(map[string]interface{}))
+	}
+	if v, ok := d.GetOk("state_notification_config"); ok {
+		deviceRegistry.StateNotificationConfig = buildStateNotificationConfig(v.(map[string]interface{}))
+	}
+	if v, ok := d.GetOk("mqtt_config"); ok {
+		deviceRegistry.MqttConfig = buildMqttConfig(v.(map[string]interface{}))
+	}
+	if v, ok := d.GetOk("http_config"); ok {
+		deviceRegistry.HttpConfig = buildHttpConfig(v.(map[string]interface{}))
+	}
+	if v, ok := d.GetOk("credentials"); ok {
+		deviceRegistry.Credentials = expandCredentials(v.([]interface{}))
+	}
+	return deviceRegistry
+}
+
+func resourceCloudIoTRegistryCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+	region, err := getRegion(d, config)
+	if err != nil {
+		return err
+	}
+	parent := fmt.Sprintf("projects/%s/locations/%s", project, region)
+
+	deviceRegistry := createDeviceRegistry(d)
+	deviceRegistry.Id = d.Get("name").(string)
+
+	call := config.clientCloudIoT.Projects.Locations.Registries.Create(parent, deviceRegistry)
+	res, err := call.Do()
+	if err != nil {
+		return err
+	}
+	d.SetId(res.Name)
```

Comment:
It can be helpful to do d.SetId(name) before the create call, then do d.SetId(nil) in the event of an error - that way, if the process is killed during the request, the state file will include the resource, and we'll be able to do a Read() on refresh to determine if the call ever actually succeeded.

Reply:
Done.